src/lfortran/semantics/ast_common_visitor.h

```diff
@@ -907,13 +907,60 @@ inline static void visit_BoolOp(Allocator &al, const AST::BoolOp_t &x,
         }) ) {
         overloaded = ASRUtils::EXPR(asr);
     }
+    // the original fails when I try for int .and. logical type
+    // and there can be logical variable, or logical pointer
+    // thus, the following fuction will unwrap it
+    auto unwrap_type = [&](ASR::ttype_t *t) -> ASR::ttype_t* {
+        // remove allocatable and pointer wrappers
+        while (ASRUtils::is_allocatable(t) || ASRUtils::is_pointer(t)) {
+            t = ASRUtils::type_get_past_allocatable(t);
+            t = ASRUtils::type_get_past_pointer(t);
+        }
+        return t;
+    };
 
-    ASR::ttype_t *left_type = ASRUtils::type_get_past_allocatable(ASRUtils::expr_type(left));
-    ASR::ttype_t *right_type = ASRUtils::type_get_past_allocatable(ASRUtils::expr_type(right));
 
-    LCOMPILERS_ASSERT(
-        ASRUtils::check_equal_type(ASRUtils::expr_type(left), ASRUtils::expr_type(right), left, right));
+    ASR::ttype_t *left_type  = unwrap_type(ASRUtils::expr_type(left));
+    ASR::ttype_t *right_type = unwrap_type(ASRUtils::expr_type(right));
 
+    auto check_logical_type = [&](ASR::ttype_t *t, ASR::expr_t *expr, const std::string &side) {
+        if (ASR::is_a<ASR::Array_t>(*t)) {
+            ASR::Array_t *arr_type = ASR::down_cast<ASR::Array_t>(t);
+            ASR::ttype_t *element_type = ASRUtils::type_get_past_allocatable(
+                                            ASRUtils::type_get_past_pointer(arr_type->m_type));
+            if (!ASR::is_a<ASR::Logical_t>(*element_type)) {
+                diag.add(Diagnostic(
+                    side + " operand's array element type must be LOGICAL",
+                    Level::Error, Stage::Semantic, {Label("", {expr->base.loc})}
+                ));
+                throw SemanticAbort();
+            }
+        } else if (!ASR::is_a<ASR::Logical_t>(*t)) {
+            diag.add(Diagnostic(
+                side + " operand must be LOGICAL or LOGICAL array",
+                Level::Error, Stage::Semantic, {Label("", {expr->base.loc})}
+            ));
+            throw SemanticAbort();
+        }
+    };
+
+    bool is_intrinsic_and = (ASRUtils::is_logical(*left_type) || ASRUtils::is_logical(*right_type));
+
+    // came across a case where there could be int .and. int
+    // this is correct. Thus, if its an intrinsic and, only then, we go for that
+    if (is_intrinsic_and) {
+        check_logical_type(left_type, left, "Left");
+        check_logical_type(right_type, right, "Right");
+    }
+
+    // this ensures both operands have the same type
+    LCOMPILERS_ASSERT(
+        ASRUtils::check_equal_type(
+            ASRUtils::expr_type(left),
+            ASRUtils::expr_type(right),
+            left, right
+        )
+    );
```

Comment:
I think you are referring to the case of an overloaded `.and.` operator here. We can simplify to:

```suggestion
   if (!overloaded && !ASRUtils::check_equal_type(ASRUtils::expr_type(left), ASRUtils::expr_type(right), left, right) {
       check_logical_type(left_type, left, "Left");
       check_logical_type(right_type, right, "Right");
   }
```